// linkedlist/lib-cov/linkedlist.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['linkedlist.js']) {
  _$jscoverage['linkedlist.js'] = [];
  _$jscoverage['linkedlist.js'][1] = 0;
  _$jscoverage['linkedlist.js'][2] = 0;
  _$jscoverage['linkedlist.js'][8] = 0;
  _$jscoverage['linkedlist.js'][14] = 0;
  _$jscoverage['linkedlist.js'][20] = 0;
  _$jscoverage['linkedlist.js'][28] = 0;
  _$jscoverage['linkedlist.js'][29] = 0;
  _$jscoverage['linkedlist.js'][32] = 0;
  _$jscoverage['linkedlist.js'][33] = 0;
  _$jscoverage['linkedlist.js'][36] = 0;
  _$jscoverage['linkedlist.js'][37] = 0;
  _$jscoverage['linkedlist.js'][40] = 0;
  _$jscoverage['linkedlist.js'][41] = 0;
  _$jscoverage['linkedlist.js'][44] = 0;
  _$jscoverage['linkedlist.js'][45] = 0;
  _$jscoverage['linkedlist.js'][46] = 0;
  _$jscoverage['linkedlist.js'][47] = 0;
  _$jscoverage['linkedlist.js'][48] = 0;
  _$jscoverage['linkedlist.js'][49] = 0;
  _$jscoverage['linkedlist.js'][51] = 0;
  _$jscoverage['linkedlist.js'][54] = 0;
  _$jscoverage['linkedlist.js'][55] = 0;
  _$jscoverage['linkedlist.js'][56] = 0;
  _$jscoverage['linkedlist.js'][57] = 0;
  _$jscoverage['linkedlist.js'][59] = 0;
  _$jscoverage['linkedlist.js'][60] = 0;
  _$jscoverage['linkedlist.js'][61] = 0;
  _$jscoverage['linkedlist.js'][62] = 0;
  _$jscoverage['linkedlist.js'][64] = 0;
  _$jscoverage['linkedlist.js'][65] = 0;
  _$jscoverage['linkedlist.js'][66] = 0;
  _$jscoverage['linkedlist.js'][67] = 0;
  _$jscoverage['linkedlist.js'][68] = 0;
  _$jscoverage['linkedlist.js'][70] = 0;
  _$jscoverage['linkedlist.js'][73] = 0;
  _$jscoverage['linkedlist.js'][74] = 0;
  _$jscoverage['linkedlist.js'][75] = 0;
  _$jscoverage['linkedlist.js'][76] = 0;
  _$jscoverage['linkedlist.js'][78] = 0;
  _$jscoverage['linkedlist.js'][79] = 0;
  _$jscoverage['linkedlist.js'][80] = 0;
  _$jscoverage['linkedlist.js'][81] = 0;
  _$jscoverage['linkedlist.js'][83] = 0;
  _$jscoverage['linkedlist.js'][84] = 0;
  _$jscoverage['linkedlist.js'][85] = 0;
  _$jscoverage['linkedlist.js'][86] = 0;
  _$jscoverage['linkedlist.js'][88] = 0;
  _$jscoverage['linkedlist.js'][91] = 0;
  _$jscoverage['linkedlist.js'][92] = 0;
  _$jscoverage['linkedlist.js'][93] = 0;
  _$jscoverage['linkedlist.js'][94] = 0;
  _$jscoverage['linkedlist.js'][95] = 0;
  _$jscoverage['linkedlist.js'][97] = 0;
  _$jscoverage['linkedlist.js'][100] = 0;
  _$jscoverage['linkedlist.js'][101] = 0;
  _$jscoverage['linkedlist.js'][102] = 0;
  _$jscoverage['linkedlist.js'][103] = 0;
  _$jscoverage['linkedlist.js'][106] = 0;
  _$jscoverage['linkedlist.js'][107] = 0;
  _$jscoverage['linkedlist.js'][108] = 0;
  _$jscoverage['linkedlist.js'][109] = 0;
  _$jscoverage['linkedlist.js'][111] = 0;
  _$jscoverage['linkedlist.js'][112] = 0;
  _$jscoverage['linkedlist.js'][113] = 0;
  _$jscoverage['linkedlist.js'][115] = 0;
  _$jscoverage['linkedlist.js'][117] = 0;
  _$jscoverage['linkedlist.js'][118] = 0;
  _$jscoverage['linkedlist.js'][119] = 0;
  _$jscoverage['linkedlist.js'][120] = 0;
  _$jscoverage['linkedlist.js'][121] = 0;
  _$jscoverage['linkedlist.js'][124] = 0;
  _$jscoverage['linkedlist.js'][125] = 0;
  _$jscoverage['linkedlist.js'][126] = 0;
  _$jscoverage['linkedlist.js'][127] = 0;
  _$jscoverage['linkedlist.js'][129] = 0;
  _$jscoverage['linkedlist.js'][130] = 0;
  _$jscoverage['linkedlist.js'][131] = 0;
  _$jscoverage['linkedlist.js'][132] = 0;
  _$jscoverage['linkedlist.js'][134] = 0;
  _$jscoverage['linkedlist.js'][135] = 0;
  _$jscoverage['linkedlist.js'][136] = 0;
  _$jscoverage['linkedlist.js'][137] = 0;
  _$jscoverage['linkedlist.js'][138] = 0;
  _$jscoverage['linkedlist.js'][139] = 0;
  _$jscoverage['linkedlist.js'][140] = 0;
  _$jscoverage['linkedlist.js'][141] = 0;
  _$jscoverage['linkedlist.js'][143] = 0;
  _$jscoverage['linkedlist.js'][144] = 0;
  _$jscoverage['linkedlist.js'][145] = 0;
  _$jscoverage['linkedlist.js'][147] = 0;
  _$jscoverage['linkedlist.js'][148] = 0;
  _$jscoverage['linkedlist.js'][151] = 0;
  _$jscoverage['linkedlist.js'][152] = 0;
  _$jscoverage['linkedlist.js'][153] = 0;
  _$jscoverage['linkedlist.js'][154] = 0;
  _$jscoverage['linkedlist.js'][155] = 0;
  _$jscoverage['linkedlist.js'][156] = 0;
  _$jscoverage['linkedlist.js'][160] = 0;
  _$jscoverage['linkedlist.js'][161] = 0;
  _$jscoverage['linkedlist.js'][162] = 0;
  _$jscoverage['linkedlist.js'][165] = 0;
  _$jscoverage['linkedlist.js'][166] = 0;
  _$jscoverage['linkedlist.js'][167] = 0;
  _$jscoverage['linkedlist.js'][168] = 0;
  _$jscoverage['linkedlist.js'][169] = 0;
  _$jscoverage['linkedlist.js'][170] = 0;
}
_$jscoverage['linkedlist.js'][1]++;
module.exports = (function () {
  _$jscoverage['linkedlist.js'][2]++;
  Object.defineProperty(this, "_head", {value: undefined, writable: true, enumerable: false, configurable: false});
  _$jscoverage['linkedlist.js'][8]++;
  Object.defineProperty(this, "_tail", {value: undefined, writable: true, enumerable: false, configurable: false});
  _$jscoverage['linkedlist.js'][14]++;
  Object.defineProperty(this, "_next", {value: undefined, writable: true, enumerable: false, configurable: false});
  _$jscoverage['linkedlist.js'][20]++;
  Object.defineProperty(this, "_length", {value: 0, writable: true, enumerable: false, configurable: false});
});
_$jscoverage['linkedlist.js'][28]++;
module.exports.prototype.__defineGetter__("head", (function () {
  _$jscoverage['linkedlist.js'][29]++;
  return this._head && this._head.data;
}));
_$jscoverage['linkedlist.js'][32]++;
module.exports.prototype.__defineGetter__("tail", (function () {
  _$jscoverage['linkedlist.js'][33]++;
  return this._tail && this._tail.data;
}));
_$jscoverage['linkedlist.js'][36]++;
module.exports.prototype.__defineGetter__("current", (function () {
  _$jscoverage['linkedlist.js'][37]++;
  return this._current && this._current.data;
}));
_$jscoverage['linkedlist.js'][40]++;
module.exports.prototype.__defineGetter__("length", (function () {
  _$jscoverage['linkedlist.js'][41]++;
  return this._length;
}));
_$jscoverage['linkedlist.js'][44]++;
module.exports.prototype.push = (function (data) {
  _$jscoverage['linkedlist.js'][45]++;
  this._tail = new Item(data, this._tail);
  _$jscoverage['linkedlist.js'][46]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][47]++;
    this._head = this._tail;
    _$jscoverage['linkedlist.js'][48]++;
    this._current = this._head;
    _$jscoverage['linkedlist.js'][49]++;
    this._next = this._head;
  }
  _$jscoverage['linkedlist.js'][51]++;
  this._length++;
});
_$jscoverage['linkedlist.js'][54]++;
module.exports.prototype.pop = (function () {
  _$jscoverage['linkedlist.js'][55]++;
  var tail = this._tail;
  _$jscoverage['linkedlist.js'][56]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][57]++;
    return;
  }
  _$jscoverage['linkedlist.js'][59]++;
  this._length--;
  _$jscoverage['linkedlist.js'][60]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][61]++;
    this._head = this._tail = this._current = this._next = undefined;
    _$jscoverage['linkedlist.js'][62]++;
    return tail.data;
  }
  _$jscoverage['linkedlist.js'][64]++;
  this._tail = tail.prev;
  _$jscoverage['linkedlist.js'][65]++;
  this._tail.next = undefined;
  _$jscoverage['linkedlist.js'][66]++;
  if (this._current === tail) {
    _$jscoverage['linkedlist.js'][67]++;
    this._current = this._tail;
    _$jscoverage['linkedlist.js'][68]++;
    this._next = undefined;
  }
  _$jscoverage['linkedlist.js'][70]++;
  return tail.data;
});
_$jscoverage['linkedlist.js'][73]++;
module.exports.prototype.shift = (function () {
  _$jscoverage['linkedlist.js'][74]++;
  var head = this._head;
  _$jscoverage['linkedlist.js'][75]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][76]++;
    return;
  }
  _$jscoverage['linkedlist.js'][78]++;
  this._length--;
  _$jscoverage['linkedlist.js'][79]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][80]++;
    this._head = this._tail = this._current = this._next = undefined;
    _$jscoverage['linkedlist.js'][81]++;
    return head.data;
  }
  _$jscoverage['linkedlist.js'][83]++;
  this._head = this._head.next;
  _$jscoverage['linkedlist.js'][84]++;
  if (this._current === head) {
    _$jscoverage['linkedlist.js'][85]++;
    this._current = this._head;
    _$jscoverage['linkedlist.js'][86]++;
    this._next = this._current.next;
  }
  _$jscoverage['linkedlist.js'][88]++;
  return head.data;
});
_$jscoverage['linkedlist.js'][91]++;
module.exports.prototype.unshift = (function (data) {
  _$jscoverage['linkedlist.js'][92]++;
  this._head = new Item(data, undefined, this._head);
  _$jscoverage['linkedlist.js'][93]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][94]++;
    this._tail = this._head;
    _$jscoverage['linkedlist.js'][95]++;
    this._next = this._head;
  }
  _$jscoverage['linkedlist.js'][97]++;
  this._length++;
});
_$jscoverage['linkedlist.js'][100]++;
module.exports.prototype.unshiftCurrent = (function () {
  _$jscoverage['linkedlist.js'][101]++;
  var current = this._current;
  _$jscoverage['linkedlist.js'][102]++;
  if (current === this._head || this._length < 2) {
    _$jscoverage['linkedlist.js'][103]++;
    return current && current.data;
  }
  _$jscoverage['linkedlist.js'][106]++;
  if (current === this._tail) {
    _$jscoverage['linkedlist.js'][107]++;
    this._tail = current.prev;
    _$jscoverage['linkedlist.js'][108]++;
    this._tail.next = undefined;
    _$jscoverage['linkedlist.js'][109]++;
    this._current = this._tail;
  }
  else {
    _$jscoverage['linkedlist.js'][111]++;
    current.next.prev = current.prev;
    _$jscoverage['linkedlist.js'][112]++;
    current.prev.next = current.next;
    _$jscoverage['linkedlist.js'][113]++;
    this._current = current.prev;
  }
  _$jscoverage['linkedlist.js'][115]++;
  this._next = this._current.next;
  _$jscoverage['linkedlist.js'][117]++;
  current.next = this._head;
  _$jscoverage['linkedlist.js'][118]++;
  current.prev = undefined;
  _$jscoverage['linkedlist.js'][119]++;
  this._head.prev = current;
  _$jscoverage['linkedlist.js'][120]++;
  this._head = current;
  _$jscoverage['linkedlist.js'][121]++;
  return current.data;
});
_$jscoverage['linkedlist.js'][124]++;
module.exports.prototype.removeCurrent = (function (data) {
  _$jscoverage['linkedlist.js'][125]++;
  var current = this._current;
  _$jscoverage['linkedlist.js'][126]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][127]++;
    return;
  }
  _$jscoverage['linkedlist.js'][129]++;
  this._length--;
  _$jscoverage['linkedlist.js'][130]++;
  if (this._length === 0) {
    _$jscoverage['linkedlist.js'][131]++;
    this._head = this._tail = this._current = this._next = undefined;
    _$jscoverage['linkedlist.js'][132]++;
    return current.data;
  }
  _$jscoverage['linkedlist.js'][134]++;
  if (current === this._tail) {
    _$jscoverage['linkedlist.js'][135]++;
    this._tail = current.prev;
    _$jscoverage['linkedlist.js'][136]++;
    this._tail.next = undefined;
    _$jscoverage['linkedlist.js'][137]++;
    this._current = this._tail;
  }
  else {
    _$jscoverage['linkedlist.js'][138]++;
    if (current === this._head) {
      _$jscoverage['linkedlist.js'][139]++;
      this._head = current.next;
      _$jscoverage['linkedlist.js'][140]++;
      this._head.prev = undefined;
      _$jscoverage['linkedlist.js'][141]++;
      this._current = this._head;
    }
    else {
      _$jscoverage['linkedlist.js'][143]++;
      current.next.prev = current.prev;
      _$jscoverage['linkedlist.js'][144]++;
      current.prev.next = current.next;
      _$jscoverage['linkedlist.js'][145]++;
      this._current = current.prev;
    }
  }
  _$jscoverage['linkedlist.js'][147]++;
  this._next = this._current.next;
  _$jscoverage['linkedlist.js'][148]++;
  return current.data;
});
_$jscoverage['linkedlist.js'][151]++;
module.exports.prototype.next = (function () {
  _$jscoverage['linkedlist.js'][152]++;
  var next = this._next;
  _$jscoverage['linkedlist.js'][153]++;
  if (next !== undefined) {
    _$jscoverage['linkedlist.js'][154]++;
    this._next = next.next;
    _$jscoverage['linkedlist.js'][155]++;
    this._current = next;
    _$jscoverage['linkedlist.js'][156]++;
    return next.data;
  }
});
_$jscoverage['linkedlist.js'][160]++;
module.exports.prototype.resetCursor = (function () {
  _$jscoverage['linkedlist.js'][161]++;
  this._current = this._next = this._head;
  _$jscoverage['linkedlist.js'][162]++;
  return this;
});
_$jscoverage['linkedlist.js'][165]++;
function Item(data, prev, next) {
  _$jscoverage['linkedlist.js'][166]++;
  this.next = next;
  _$jscoverage['linkedlist.js'][167]++;
  if (next) {
    _$jscoverage['linkedlist.js'][167]++;
    next.prev = this;
  }
  _$jscoverage['linkedlist.js'][168]++;
  this.prev = prev;
  _$jscoverage['linkedlist.js'][169]++;
  if (prev) {
    _$jscoverage['linkedlist.js'][169]++;
    prev.next = this;
  }
  _$jscoverage['linkedlist.js'][170]++;
  this.data = data;
}
_$jscoverage['linkedlist.js'].source = ["module.exports = function () {","  Object.defineProperty(this, '_head', {","    value: undefined,","    writable: true,","    enumerable: false,","    configurable: false","  })","  Object.defineProperty(this, '_tail', {","    value: undefined,","    writable: true,","    enumerable: false,","    configurable: false","  })","  Object.defineProperty(this, '_next', {","    value: undefined,","    writable: true,","    enumerable: false,","    configurable: false","  })","  Object.defineProperty(this, '_length', {","    value: 0,","    writable: true,","    enumerable: false,","    configurable: false","  })","}","","module.exports.prototype.__defineGetter__('head', function () {","  return this._head &amp;&amp; this._head.data","})","","module.exports.prototype.__defineGetter__('tail', function () {","  return this._tail &amp;&amp; this._tail.data","})","","module.exports.prototype.__defineGetter__('current', function () {","  return this._current &amp;&amp; this._current.data","})","","module.exports.prototype.__defineGetter__('length', function () {","  return this._length","})","","module.exports.prototype.push = function (data) {","  this._tail = new Item(data, this._tail)","  if (this._length === 0) {","    this._head = this._tail","    this._current = this._head","    this._next = this._head","  }","  this._length++","}","","module.exports.prototype.pop = function () {","  var tail = this._tail","  if (this._length === 0) {","    return","  }","  this._length--","  if (this._length === 0) {","    this._head = this._tail = this._current = this._next = undefined","    return tail.data","  }","  this._tail = tail.prev","  this._tail.next = undefined","  if (this._current === tail) {","    this._current = this._tail","    this._next = undefined","  }","  return tail.data","}","","module.exports.prototype.shift = function () {","  var head = this._head","  if (this._length === 0)  {","    return","  }","  this._length--","  if (this._length === 0) {","    this._head = this._tail = this._current = this._next = undefined","    return head.data","  }","  this._head = this._head.next","  if (this._current === head) {","    this._current = this._head","    this._next = this._current.next","  }","  return head.data","}","","module.exports.prototype.unshift = function (data) {","  this._head = new Item(data, undefined, this._head)","  if (this._length === 0)  {","    this._tail = this._head","    this._next = this._head","  }","  this._length++","}","","module.exports.prototype.unshiftCurrent = function () {","  var current = this._current","  if (current === this._head || this._length &lt; 2) {","    return current &amp;&amp; current.data","  }","  // remove","  if (current === this._tail) {","    this._tail = current.prev","    this._tail.next = undefined","    this._current = this._tail","  } else {","    current.next.prev = current.prev","    current.prev.next = current.next","    this._current = current.prev","  }","  this._next = this._current.next","  // unshift","  current.next = this._head","  current.prev = undefined","  this._head.prev = current","  this._head = current","  return current.data","}","","module.exports.prototype.removeCurrent = function (data) {","  var current = this._current","  if (this._length === 0) {","    return","  }","  this._length--","  if (this._length === 0) {","    this._head = this._tail = this._current = this._next = undefined","    return current.data","  }","  if (current === this._tail) {","    this._tail = current.prev","    this._tail.next = undefined","    this._current = this._tail","  } else if (current === this._head) {","    this._head = current.next","    this._head.prev = undefined","    this._current = this._head","  } else {","    current.next.prev = current.prev","    current.prev.next = current.next","    this._current = current.prev","  }","  this._next = this._current.next","  return current.data","}","","module.exports.prototype.next = function () {","  var next = this._next","  if (next !== undefined) {","    this._next = next.next","    this._current = next","    return next.data","  }","}","","module.exports.prototype.resetCursor = function () {","  this._current = this._next = this._head","  return this","}","","function Item (data, prev, next) {","  this.next = next","  if (next) next.prev = this","  this.prev = prev","  if (prev) prev.next = this","  this.data = data","}"];
